compile/compile.go

```diff
@@ -0,0 +1,603 @@
+	result.Data = load.Files.Documents
+
+	var modules []string
+
+	for k := range load.Files.Modules {
+		modules = append(modules, k)
+	}
+
+	sort.Strings(modules)
+
+	for _, module := range modules {
+		result.Modules = append(result.Modules, bundle.ModuleFile{
+			URL:    load.Files.Modules[module].Name,
+			Path:   load.Files.Modules[module].Name,
+			Parsed: load.Files.Modules[module].Parsed,
+			Raw:    load.Files.Modules[module].Raw,
+		})
+	}
+
+	c.bundle = result
+
+	return nil
+}
+
+func (c *Compiler) optimize(ctx context.Context) error {
+
+	if c.optimizationLevel <= 0 {
+		var err error
+		c.compiler, err = compile(c.bundle)
+		return err
+	}
+
+	o := newOptimizer(c.bundle).
+		WithEntrypoints(c.entrypointrefs).
+		WithDebug(c.debug)
+
+	err := o.Do(ctx)
+	if err != nil {
+		return err
+	}
+
+	// NOTE(tsandall): do not use the compiler from the optimizer as it will
+	// be stale and not reflect the output bundle contents.
+	c.bundle = o.bundle
+
+	return nil
+}
+
+func (c *Compiler) compileWasm(ctx context.Context) error {
+
+	if c.compiler == nil {
+		var err error
+		c.compiler, err = compile(c.bundle)
+		if err != nil {
+			return err
+		}
+	}
+
+	store := inmem.NewFromObject(c.bundle.Data)
+	resultSym := ast.VarTerm(ast.WildcardPrefix + "result")
+
+	cr, err := rego.New(
+		rego.ParsedQuery(ast.NewBody(ast.Equality.Expr(resultSym, c.entrypointrefs[0]))),
+		rego.Compiler(c.compiler),
+		rego.Store(store),
+	).Compile(ctx)
+
+	if err != nil {
+		return err
+	}
+
+	c.bundle.Wasm = cr.Bytes
+
+	return nil
+}
+
+type undefinedEntrypointErr struct {
+	Entrypoint *ast.Term
+}
+
+func (err undefinedEntrypointErr) Error() string {
+	return fmt.Sprintf("undefined entrypoint %v", err.Entrypoint)
+}
+
+type optimizer struct {
+	bundle          *bundle.Bundle
+	compiler        *ast.Compiler
+	entrypoints     []*ast.Term
+	nsprefix        string
+	resultsymprefix string
+	outputprefix    string
+	debug           *debugEvents
+}
+
+func newOptimizer(b *bundle.Bundle) *optimizer {
+	return &optimizer{
+		bundle:          b,
+		nsprefix:        "partial",
+		resultsymprefix: ast.WildcardPrefix,
+		outputprefix:    "optimized",
+	}
+}
+
+func (o *optimizer) WithDebug(debug *debugEvents) *optimizer {
+	o.debug = debug
+	return o
+}
+
+func (o *optimizer) WithEntrypoints(es []*ast.Term) *optimizer {
+	o.entrypoints = es
+	return o
+}
+
+func (o *optimizer) Do(ctx context.Context) error {
+
+	// TODO(tsandall): implement optimization levels. These will just be params on partial evaluation for now.
+	//
+	// Level 1: PE w/ constant folding. Only inline rules that are completely known.
+	// Level 2: L1 except inlining of rules with unknowns.
+	// Level 3: L2 except aggressive inlining using negation and copy propagation optimizations.
+
+	// NOTE(tsandall): if there are multiple entrypoints, copy the bundle because
+	// if any of the optimization steps fail, we do not want to leave the caller's
+	// bundle in a partially modified state.
```

Comment:
I don't have a strong opinion here. Since the caller is internal only for now, I'd be inclined to just leave it and avoid the unnecessary deep copy of all of the modules and all of the data.